ports/stm32f4/peripherals/stm32f4/pins.h

```diff
@@ -38,28 +38,32 @@
 
 typedef struct {
     mp_obj_base_t base;
-    //uint8_t number; //(3)port,(5)pin 
-    uint8_t port_number;
-    GPIO_TypeDef * port;
     uint8_t number;
+    uint8_t adc_num_input; //(3)mask ()
```

Comment:
Instead of doing the bitpacking yourself please use the `: <no bits>;` form when defining the struct like we use here: https://github.com/adafruit/samd-peripherals/blob/83a4759d186574d8034435cd2303def85e4ed793/samd/pins.h#L46 That way the compiler will handle all of the packing and unpacking for you and the source will be more readable.

For the ADC1, ADC2, ADC3 I'd define a macro for each and then | them together for the combinations 12 and 23.

Reply:
Regarding the ADCs: the reason I had distinct macros is that those are the only ADC combinations possible on STM32 (as far as I've seen across the whole F4 line). I felt it'd be helpful context to call that out, since whole banks of pins tend to be grouped into one of the categories. Do you think individual ADC triggers would still be better?

Reply:
Ah ok! Please add a comment stating that reasoning next to the macros. Makes total sense.